File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque 
def solution(maps): 
    dx=[-1,1,0,0] 
    dy=[0,0,-1,1]

    def bfs (x,y):
        queue=deque()
        queue.append((x,y))
        while queue: 
            x,y=queue.popleft() 
            for i in range(4):
                nx=x+dx[i]
                ny=y+dy[i]
                if 0>nx or 0>ny or nx>=len(maps) or ny>=len(maps[0]):
                    continue
                if maps[nx][ny]==0: 
                    continue
                if maps[nx][ny]==1: 
                    queue.append((nx,ny)) 
                    maps[nx][ny]=maps[x][y]+1  
    bfs(0,0)
    answer=maps[len(maps)-1][len(maps[0])-1]

    if answer==1: 
        answer=-1
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    public int solution(int[][] maps) {
        
        return bfs(maps);
    }
    public int bfs(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        int[][] d = new int[][]{{1,0}, {-1,0}, {0,1}, {0,-1}};
        int[][] visited = new int[n][m];
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0, 1});
        visited[0][0] = 1;
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            for (int[] move : d) {
                int ny = curr[0] + move[0];
                int nx = curr[1] + move[1];
                if (0 > ny || ny >= n || 0 > nx || nx >= m || maps[ny][nx] == 0 || visited[ny][nx] != 0) {
                    continue;
                }
                visited[ny][nx] = curr[2] + 1;
                q.add(new int[]{ny, nx, curr[2]+1});
                if (ny == n-1 && nx == m-1) {
                    return visited[ny][nx];
                }
            }
        }
        return -1;
    }
}